lib/test_prof/memory_prof/tracker/linked_list.rb

```diff
@@ -0,0 +1,111 @@
+      # current examples/groups, then the head node will always point to a leaf in that tree.
+      #
+      # For example, if we have the following spec:
+      #
+      #   describe Question do
+      #     decribe "#publish" do
+      #       context "when not published" do
+      #         it "makes the question visible" do
+      #           ...
+      #         end
+      #       end
+      #     end
+      #   end
+      #
+      # At the moment when rspec is executing the example, the list has the following structure
+      # (^ denotes the head node):
+      #
+      #    ^"makes the question visible" ->  "when not published" -> "#publish" -> Question
+      #
+      # LinkedList supports two method for working with it:
+      #
+      #  * add_node – adds a node to the beginig of the list. At this point an example or group
+      #    has started and we track how much memory has already been used.
+      #  * remove_node – removes and returns the head node from the list. It means that the node
+      #    example/group has finished and it is time to calculate its memory usage.
+      #
+      # When we remove a node we add its total_memory to the previous node.nested_memory, thus
+      # gradually tracking the amount of memory used by nested examples inside a group.
+      #
+      # In the example above, after we remove the node "makes the question visible", we add its total
+      # memory usage to nested_memory of the "when not published" node. If the "when not published"
+      # group contains other examples or sub-groups, their total_memory will also be added to
+      # "when not published" nested_memory. So when the group finishes we will have the total amount
+      # of memory used by its nested examples/groups, and thus we will be able to calculate the memory
+      # used by hooks and other code inside a group by subtracting nested_memory from total_memory.
+      class LinkedList
+        attr_reader :head
+
+        def initialize(memory_at_start)
+          add_node(:total, memory_at_start)
+        end
+
+        def add_node(item, memory_at_start)
+          @head = LinkedListNode.new(
+            item: item,
+            previous: head,
+            memory_at_start: memory_at_start
+          )
+        end
+
+        def remove_node(item, memory_at_finish)
+          return if head.item != item
+          head.finish(memory_at_finish)
+
+          current = head
+          @head = head.previous
+
+          current
+        end
+      end
+
+      class LinkedListNode
+        attr_reader :item, :previous, :memory_at_start, :memory_at_finish, :nested_memory
+
+        def initialize(item:, memory_at_start:, previous:)
+          @item = item
+          @previous = previous
+
+          @memory_at_start = memory_at_start
+          @memory_at_finish = nil
+          @nested_memory = 0
+        end
+
+        def total_memory
+          memory_at_finish - memory_at_start
```

Comment:
`memory_at_finish` can potentially be `nil` before the method `finish` is called.

Maybe it makes sense to handle this case through some exception?